// apify/configuration.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Configuration = void 0;
const consts_1 = require("@apify/consts");
const core_1 = require("@crawlee/core");
/**
 * `Configuration` is a value object holding the SDK configuration. We can use it in two ways:
 *
 * 1. When using `Actor` class, we can get the instance configuration via `sdk.config`
 *
 *    ```javascript
 *    import { Actor } from 'apify';
 *    import { BasicCrawler } from 'crawlee';
 *
 *    const sdk = new Actor({ token: '123' });
 *    console.log(sdk.config.get('token')); // '123'
 *
 *    const crawler = new BasicCrawler({
 *        // ... crawler options
 *    }, sdk.config);
 *    ```
 *
 * 2. To get the global configuration (singleton instance). It will respect the environment variables.
 *
 *    ```javascript
 *    import { BasicCrawler, Configuration } from 'crawlee';
 *
 *    const config = new Configuration({ persistStateIntervalMillis: 30_000 });
 *    const crawler = new BasicCrawler({ ... }, config);
 *    ```
 *
 * ## Supported Configuration Options
 *
 * Key | Environment Variable | Default Value
 * ---|---|---
 * `memoryMbytes` | `APIFY_MEMORY_MBYTES` | -
 * `headless` | `APIFY_HEADLESS` | -
 * `persistStateIntervalMillis` | `APIFY_PERSIST_STATE_INTERVAL_MILLIS` | `60e3`
 * `token` | `APIFY_TOKEN` | -
 * `isAtHome` | `APIFY_IS_AT_HOME` | -
 * `defaultDatasetId` | `APIFY_DEFAULT_DATASET_ID` | `'default'`
 * `defaultKeyValueStoreId` | `APIFY_DEFAULT_KEY_VALUE_STORE_ID` | `'default'`
 * `defaultRequestQueueId` | `APIFY_DEFAULT_REQUEST_QUEUE_ID` | `'default'`
 *
 * ## Advanced Configuration Options
 *
 * Key | Environment Variable | Default Value
 * ---|---|---
 * `actorEventsWsUrl` | `APIFY_ACTOR_EVENTS_WS_URL` | -
 * `actorId` | `APIFY_ACTOR_ID` | -
 * `actorRunId` | `APIFY_ACTOR_RUN_ID` | -
 * `actorTaskId` | `APIFY_ACTOR_TASK_ID` | -
 * `apiBaseUrl` | `APIFY_API_BASE_URL` | `'https://api.apify.com'`
 * `containerPort` | `APIFY_CONTAINER_PORT` | `4321`
 * `containerUrl` | `APIFY_CONTAINER_URL` | `'http://localhost:4321'`
 * `inputKey` | `APIFY_INPUT_KEY` | `'INPUT'`
 * `metamorphAfterSleepMillis` | `APIFY_METAMORPH_AFTER_SLEEP_MILLIS` | `300e3`
 * `proxyHostname` | `APIFY_PROXY_HOSTNAME` | `'proxy.apify.com'`
 * `proxyPassword` | `APIFY_PROXY_PASSWORD` | -
 * `proxyPort` | `APIFY_PROXY_PORT` | `8000`
 * `proxyStatusUrl` | `APIFY_PROXY_STATUS_URL` | `'http://proxy.apify.com'`
 * `userId` | `APIFY_USER_ID` | -
 * `xvfb` | `APIFY_XVFB` | -
 * `chromeExecutablePath` | `APIFY_CHROME_EXECUTABLE_PATH` | -
 * `defaultBrowserPath` | `APIFY_DEFAULT_BROWSER_PATH` | -
 */
class Configuration extends (_b = core_1.Configuration) {
    /**
     * @inheritDoc
     */
    get(key, defaultValue) {
        return super.get(key, defaultValue);
    }
    /**
     * @inheritDoc
     */
    set(key, value) {
        super.set(key, value);
    }
    /**
     * @inheritDoc
     */
    static getGlobalConfig() {
        if (Configuration.storage.getStore()) {
            return Configuration.storage.getStore();
        }
        Configuration.globalConfig ?? (Configuration.globalConfig = new Configuration());
        return Configuration.globalConfig;
    }
    /**
     * Resets global configuration instance. The default instance holds configuration based on env vars,
     * if we want to change them, we need to first reset the global state. Used mainly for testing purposes.
     */
    static resetGlobalState() {
        delete this.globalConfig;
    }
}
exports.Configuration = Configuration;
_a = Configuration;
// maps environment variables to config keys (e.g. `APIFY_MEMORY_MBYTES` to `memoryMbytes`)
Object.defineProperty(Configuration, "ENV_MAP", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: {
        // regular crawlee env vars are also supported
        ...Reflect.get(_b, "ENV_MAP", _a),
        // support crawlee env vars prefixed with `APIFY_` too
        APIFY_AVAILABLE_MEMORY_RATIO: 'availableMemoryRatio',
        APIFY_PURGE_ON_START: 'purgeOnStart',
        APIFY_MEMORY_MBYTES: 'memoryMbytes',
        APIFY_DEFAULT_DATASET_ID: 'defaultDatasetId',
        APIFY_DEFAULT_KEY_VALUE_STORE_ID: 'defaultKeyValueStoreId',
        APIFY_DEFAULT_REQUEST_QUEUE_ID: 'defaultRequestQueueId',
        APIFY_INPUT_KEY: 'inputKey',
        APIFY_PERSIST_STATE_INTERVAL_MILLIS: 'persistStateIntervalMillis',
        APIFY_HEADLESS: 'headless',
        APIFY_XVFB: 'xvfb',
        APIFY_CHROME_EXECUTABLE_PATH: 'chromeExecutablePath',
        APIFY_DEFAULT_BROWSER_PATH: 'defaultBrowserPath',
        APIFY_DISABLE_BROWSER_SANDBOX: 'disableBrowserSandbox',
        // as well as apify specific ones
        APIFY_TOKEN: 'token',
        APIFY_METAMORPH_AFTER_SLEEP_MILLIS: 'metamorphAfterSleepMillis',
        APIFY_TEST_PERSIST_INTERVAL_MILLIS: 'persistStateIntervalMillis',
        APIFY_ACTOR_EVENTS_WS_URL: 'actorEventsWsUrl',
        APIFY_ACTOR_ID: 'actorId',
        APIFY_API_BASE_URL: 'apiBaseUrl',
        APIFY_IS_AT_HOME: 'isAtHome',
        APIFY_ACTOR_RUN_ID: 'actorRunId',
        APIFY_ACTOR_TASK_ID: 'actorTaskId',
        APIFY_CONTAINER_PORT: 'containerPort',
        APIFY_CONTAINER_URL: 'containerUrl',
        APIFY_USER_ID: 'userId',
        APIFY_PROXY_HOSTNAME: 'proxyHostname',
        APIFY_PROXY_PASSWORD: 'proxyPassword',
        APIFY_PROXY_STATUS_URL: 'proxyStatusUrl',
        APIFY_PROXY_PORT: 'proxyPort',
    }
});
Object.defineProperty(Configuration, "INTEGER_VARS", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: [...Reflect.get(_b, "INTEGER_VARS", _a), 'proxyPort', 'containerPort', 'metamorphAfterSleepMillis']
});
Object.defineProperty(Configuration, "BOOLEAN_VARS", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: [...Reflect.get(_b, "BOOLEAN_VARS", _a), 'isAtHome']
});
Object.defineProperty(Configuration, "DEFAULTS", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: {
        ...Reflect.get(_b, "DEFAULTS", _a),
        defaultKeyValueStoreId: consts_1.LOCAL_ENV_VARS[consts_1.ENV_VARS.DEFAULT_KEY_VALUE_STORE_ID],
        defaultDatasetId: consts_1.LOCAL_ENV_VARS[consts_1.ENV_VARS.DEFAULT_DATASET_ID],
        defaultRequestQueueId: consts_1.LOCAL_ENV_VARS[consts_1.ENV_VARS.DEFAULT_REQUEST_QUEUE_ID],
        inputKey: 'INPUT',
        apiBaseUrl: 'https://api.apify.com',
        proxyStatusUrl: 'http://proxy.apify.com',
        proxyHostname: consts_1.LOCAL_ENV_VARS[consts_1.ENV_VARS.PROXY_HOSTNAME],
        proxyPort: +consts_1.LOCAL_ENV_VARS[consts_1.ENV_VARS.PROXY_PORT],
        containerPort: +consts_1.LOCAL_ENV_VARS[consts_1.ENV_VARS.CONTAINER_PORT],
        containerUrl: consts_1.LOCAL_ENV_VARS[consts_1.ENV_VARS.CONTAINER_URL],
        metamorphAfterSleepMillis: 300e3,
        persistStateIntervalMillis: 60e3, // This value is mentioned in jsdoc in `events.js`, if you update it here, update it there too.
    }
});
// monkey patch the core class so it respects the new options too
core_1.Configuration.getGlobalConfig = Configuration.getGlobalConfig;
// @ts-expect-error protected property
core_1.Configuration.ENV_MAP = Configuration.ENV_MAP;
// @ts-expect-error protected property
core_1.Configuration.INTEGER_VARS = Configuration.INTEGER_VARS;
// @ts-expect-error protected property
core_1.Configuration.BOOLEAN_VARS = Configuration.BOOLEAN_VARS;
// @ts-expect-error protected property
core_1.Configuration.DEFAULTS = Configuration.DEFAULTS;
